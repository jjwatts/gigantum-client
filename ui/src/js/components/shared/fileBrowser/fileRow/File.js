// vendor
import React, { Component } from 'react';
import Moment from 'moment';
import fileIconsJs from 'file-icons-js';
import classNames from 'classnames';
import { DragSource, DropTarget } from 'react-dnd';
import { NativeTypes } from 'react-dnd-html5-backend';
import TextTruncate from 'react-text-truncate';
// config
import config from 'JS/config';
// store
import store from 'JS/redux/store';
import { setMergeMode, updateTransitionState } from 'JS/redux/actions/labbook/labbook';
import { setErrorMessage, setInfoMessage, setWarningMessage } from 'JS/redux/actions/footer';
// mutations
import StartContainerMutation from 'Mutations/container/StartContainerMutation';
import StartDevToolMutation from 'Mutations/container/StartDevToolMutation';
// components
import ActionsMenu from './ActionsMenu';
import DatasetActionsMenu from './dataset/DatasetActionsMenu';
// utils
import Connectors from '../utilities/Connectors';
// assets
import './File.scss';

class File extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isDragging: props.isDragging,
      isSelected: (props.isSelected || (this.props.childrenState[this.props.fileData.edge.node.key] && this.props.childrenState[this.props.fileData.edge.node.key].isSelected)) || false,
      stateSwitch: false,
      newFileName: props.filename,
      renameEditMode: false,
      hover: false,
    };
    this._setSelected = this._setSelected.bind(this);
    this._renameEditMode = this._renameEditMode.bind(this);
    this._triggerMutation = this._triggerMutation.bind(this);
    this._clearState = this._clearState.bind(this);
    this._setHoverState = this._setHoverState.bind(this);
    this._checkHover = this._checkHover.bind(this);
  }

  static getDerivedStateFromProps(nextProps, state) {
    const isSelected = (nextProps.multiSelect === 'all')
      ? true
      : (nextProps.multiSelect === 'none')
        ? false
        : state.isSelected;

    if ((nextProps.isOverCurrent !== nextProps.isOverChildFile) && !nextProps.isDragging && state.hover) {
      nextProps.updateParentDropZone(nextProps.isOverCurrent);
    }
    return {
      ...state,
      isSelected,
    };
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    const { state } = this;
    if (state.renameEditMode) {
      this.reanmeInput.focus();
    }
  }

  /**
  *  @param {boolean} isSelected - sets if file has been selected
  *  sets elements to be selected and parent
  */
  _setSelected(evt, isSelected) {
    const { props } = this;
    props.updateChildState(props.fileData.edge.node.key, isSelected, false);
    this.setState({ isSelected }, () => {
      Object.keys(this.refs).forEach((ref) => {
        this.refs[ref].getDecoratedComponentInstance().getDecoratedComponentInstance()._setSelected();
      });
      if (props.checkParent) {
        props.checkParent();
      }
    });
  }

  /**
    *  @param {}
    *  sets dragging state
    */
  _mouseEnter() {
    if (this.props.setParentDragFalse) {
      this.props.setParentDragFalse();
    }
    if (this.state.isDragging && this.state.isHovered) {
      this.setState({ isDragging: true, isHovered: true });
    }
  }

  /**
    *  @param {String} devTool
    *  @param {Boolean} forceStart
    *  launches dev tool to appropriate file
    */
  _openDevTool(devTool, forceStart) {
    const { props } = this;
    const { owner, labbookName } = store.getState().routes;
    const status = props.containerStatus;
    const tabName = `${devTool}-${owner}-${labbookName}`;

    if ((status !== 'NOT_RUNNING') && (status !== 'RUNNING')) {
      setWarningMessage('Could not launch development environment as the project is not ready.');
    } else if (status === 'NOT_RUNNING' && !forceStart) {
      setInfoMessage('Starting Project container. When done working, click Stop to shutdown the container.');
      updateTransitionState(labbookName, 'Starting');
      setMergeMode(false, false);
      StartContainerMutation(
        owner,
        labbookName,
        (response, error) => {
          if (error) {
            setErrorMessage(`There was a problem starting ${labbookName} container`, error);
          } else {
            this._openDevTool(devTool, true);
          }
        },
      );
    } else {
      setInfoMessage(`Starting ${devTool}, make sure to allow popups.`);
      StartDevToolMutation(
        owner,
        labbookName,
        devTool,
        (response, error) => {
          if (response.startDevTool) {
            let path = `${window.location.protocol}//${window.location.hostname}${response.startDevTool.path}`;
            if (path.includes(`/lab/tree/${props.section}`)) {
              path = path.replace(`/lab/tree/${props.section}`, `/lab/tree/${props.section}/${props.fileData.edge.node.key}`);
            } else if (props.fileData.edge.node.key.indexOf('.Rmd') === -1) {
              path = `${path}/lab/tree/${props.section}/${props.fileData.edge.node.key}`;
            }

            window[tabName] = window.open(path, tabName);
            window[tabName].close();
            window[tabName] = window.open(path, tabName);
          }

          if (error) {
            setErrorMessage('Error Starting Dev tool', error);
          }
        },
      );
    }
  }

  /**
  *  @param {}
  *  sets dragging state
  */
  _mouseLeave() {
    const { props, state } = this;
    if (props.setParentDragTrue) {
      props.setParentDragTrue();
    }
    if (!state.isDragging && !state.isHovered) {
      this.setState({ isDragging: false, isHovered: false });
    }
  }

  /**
  *  @param {}
  *  sets dragging state to true
  */
  _checkHover() {
    const { state } = this;
    if (state.isHovered && !state.isDragging) {
      this.setState({ isDragging: true });
    }
  }

  /**
  *  @param {boolean} renameEditMode
  *  sets dragging state
  */
  _renameEditMode(renameEditMode) {
    this.setState({ renameEditMode });
  }

  /**
  *  @param {Object} evt
  *  sets dragging state
  */
  _updateFileName(evt) {
    this.setState({
      newFileName: evt.target.value,
    });
  }

  /**
  *  @param {Object} evt
  *  sets dragging state
  *  @return {}
  */
  _submitRename(evt, fileName) {
    if (evt.key === 'Enter' && this.state.newFileName !== fileName) {
      this._triggerMutation();
    }

    if (evt.key === 'Escape') {
      this._clearState();
    }
  }

  /**
  *  @param {}
  *  sets state on a boolean value
  *  @return {}
  */
  _clearState() {
    const { props } = this;
    this.setState({
      newFileName: props.filename,
      renameEditMode: false,
    });
  }

  /**
  *  @param {}
  *  triggers move mutation and clearState function
  *  @return {}
  */
  _triggerMutation() {
    const { props, state } = this;
    const fileKeyArray = props.fileData.edge.node.key.split('/');
    fileKeyArray.pop();
    const folderKeyArray = fileKeyArray;
    const folderKey = folderKeyArray.join('/');
    const newKey = (folderKey.length > 0)
      ? `${folderKey}/${state.newFileName}` : state.newFileName;

    this._clearState();
    if (props.fileData.edge.node.key !== newKey) {
      const data = {
        newKey: `${folderKey}/${state.newFileName}`,
        edge: props.fileData.edge,
        removeIds: [props.fileData.edge.node.id],
      };

      if (props.section !== 'data') {
        props.mutations.moveLabbookFile(data, (response) => {
          this._clearState();
        });
      } else {
        props.mutations.moveDatasetFile(data, (response) => {
          this._clearState();
        });
      }
    }
  }

  /**
  *  @param {object} event
  *  @param {boolean} hover - hover state for mouseover
  *  sets hover state
  *  @return {}
  */
  _setHoverState(evt, hover) {
    const { props, state } = this;
    evt.preventDefault();

    if (state.hover !== hover) {
      this.setState({ hover });
    }

    if (props.setParentHoverState && hover) {
      props.setParentHoverState(evt, !hover);
    }
  }

  render() {
    const { props, state } = this;
    const { node } = props.fileData.edge;
    const { index } = props.fileData;
    const fileName = props.filename;
    const cantDrag = !props.fileData.edge.node.isLocal && props.section === 'data'
    const isNotebook = fileName.split('.')[fileName.split('.').length - 1] === 'ipynb';
    // TODO remove hard coded references when api is updated
    const isRFile = fileName.split('.')[fileName.split('.').length - 1] === 'Rmd';
    const devTool = isNotebook ? 'JupyterLab' : isRFile ? 'Rstudio' : '';
    const paddingLeft = 40 * index;
    const rowStyle = { paddingLeft: `${paddingLeft}px` };

    // declare css here
    const fileRowCSS = classNames({
      File__row: true,
      'File__row--hover': state.hover,
      'File__row--noDrag': props.isDragging && cantDrag,
      'File__row--canDrag': props.isDragging && !cantDrag,
    });
    const buttonCSS = classNames({
      CheckboxMultiselect: true,
      CheckboxMultiselect__uncheck: !state.isSelected,
      CheckboxMultiselect__check: state.isSelected,
    });
    const textIconsCSS = classNames({
      'File__cell File__cell--name': true,
      hidden: state.renameEditMode,
    });
    const renameCSS = classNames({
      'File__cell File__cell--edit': true,
      hidden: !state.renameEditMode,
    });
    const truncateCSS = classNames({
      File__paragragh: true,
      'File__paragragh--external': isNotebook || isRFile,
    });

    const file = (
      <div
        style={this.props.style}
        onMouseOver={(evt) => { this._setHoverState(evt, true); }}
        onMouseOut={(evt) => { this._setHoverState(evt, false); }}
        onMouseLeave={() => { this._mouseLeave(); }}
        onMouseEnter={() => { this._mouseEnter(); }}
        className="File"
      >

        <div
          className={fileRowCSS}
          style={rowStyle}
        >
          {
            !props.readOnly
            && (
            <button
              type="button"
              className={buttonCSS}
              onClick={(evt) => { this._setSelected(evt, !this.state.isSelected); }}
            />
            )
          }
          <div className={textIconsCSS}>

            <div className={`File__icon ${fileIconsJs.getClass(fileName)}`} />

            <div className="File__text">
              { props.expanded
                && (
                <TextTruncate
                  className={truncateCSS}
                  line={1}
                  truncateText="…"
                  text={fileName}
                  onClick={() => {
                    if (isNotebook || isRFile) {
                      this._openDevTool(devTool);
                    }
                  }}
                />
                )
              }
            </div>
          </div>

          <div className={renameCSS}>

            <div className="File__container">
              <input
                draggable
                ref={(input) => { this.reanmeInput = input; }}
                value={this.state.newFileName}
                type="text"
                className="File__input"
                onClick={(evt) => { evt.preventDefault(); evt.stopPropagation(); }}
                onDragStart={(evt) => { evt.preventDefault(); evt.stopPropagation(); }}
                onChange={(evt) => { this._updateFileName(evt); }}
                onKeyDown={(evt) => { this._submitRename(evt, fileName); }}
              />
            </div>
            <div className="flex justify-space-around">
              <button
                type="button"
                className="File__btn--round File__btn--cancel File__btn--rename-cancel"
                onClick={() => { this._clearState(); }}
              />
              <button
                type="button"
                className="File__btn--round File__btn--add File__btn--rename-add"
                disabled={this.state.newFileName === fileName}
                onClick={() => { this._triggerMutation(); }}
              />
            </div>
          </div>

          <div className="File__cell File__cell--size">
            {config.humanFileSize(node.size)}
          </div>

          <div className="File__cell File__cell--date">
            {Moment((node.modifiedAt * 1000), 'x').fromNow()}
          </div>

          <div className="File__cell File__cell--menu">
            {
              !props.readOnly
              && (
              <ActionsMenu
                edge={props.fileData.edge}
                mutationData={props.mutationData}
                mutations={props.mutations}
                renameEditMode={this._renameEditMode}
                section={props.section}
              />
              )
            }
            {
              (props.section === 'data')
              && (
              <DatasetActionsMenu
                edge={props.fileData.edge}
                section={props.section}
                mutationData={props.mutationData}
                mutations={props.mutations}
                renameEditMode={this._renameEditMode}
                isDownloading={props.isDownloading}
                parentDownloading={props.parentDownloading}
                setFolderIsDownloading={props.setFolderIsDownloading}
                isLocal={props.fileData.edge.node.isLocal}
                isDragging={props.isDragging}
              />
              )
            }
          </div>

        </div>
      </div>
    );

    if (!props.readOnly) {
      return (
        props.connectDragSource(file)
      );
    }
    return file;
  }
}

const FileDnD = DragSource(
  'card',
  Connectors.dragSource,
  Connectors.dragCollect,
)((File));


export default FileDnD;
