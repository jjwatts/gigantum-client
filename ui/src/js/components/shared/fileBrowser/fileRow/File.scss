@import '~Styles/_imports.scss';

.UserSelect--none{
  -webkit-touch-callout: none; /* iOS Safari */
-webkit-user-select: none; /* Safari */
 -khtml-user-select: none; /* Konqueror HTML */
   -moz-user-select: none; /* Firefox */
    -ms-user-select: none; /* Internet Explorer/Edge */
        user-select: none; /* Non-prefixed version, currently
                              supported by Chrome and Opera */
}

.File__row{
    //background: $white;
    height: 50px;
    padding-top: 12px;
    padding-bottom: 12px;

    cursor: pointer;

    border-bottom: 1px solid $romanSilver;

    @include flex(space-between, row);
    align-items: center;

    @extend .UserSelect--none;

    &--noDrag {
      background: $background-color !important;
      cursor: not-allowed !important;
    }
    &--canDrag {
      cursor: copy !important;
    }

    &--hover {
      background: $fileBrowserHover;
    }
}

.File__cell{

  @include flex(flex-start, row);

  @extend .UserSelect--none;

  &--name{
      margin: 0 20px;
      flex: 1;
      @extend .UserSelect--none;
  }

  &--size{
      width: 100px;
  }

  &--date{
      width: 140px;
  }

  &--menu{
    width: 180px;
  }

  &--edit{
    @include flex(space-around, row);
    flex: 1;
    margin: 0 20px;
  }

}

.File__text{
  flex: 1;
  text-align: left;
  align-self: center;
  position: relative;
  @extend .UserSelect--none;
}

.File__icon{
    margin-right: 5px;
    @extend .UserSelect--none;
}

.File__btn{

    width: 20px;
    height: 20px;
    margin: 0 0 0 20px;
    padding: 0;

    border: none;
    background: url('~Images/icons/check-unselected.svg') no-repeat 50% 50% !important;
    background-size: 26px !important;

    &:hover {
        border: none;
        background: $azure url('~Images/icons/check-selected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
    }

    &:focus {
        border: none;
        background: url('~Images/icons/check-unselected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
    }

    &--selected {
        border: none;
        background: $azure url('~Images/icons/check-selected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
    }

    &--selected:hover {
        border: none;
        background: url('~Images/icons/check-unselected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
    }

    &--selected:focus {
        border: none;
        background: $azure url('~Images/icons/check-selected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
    }

    &--incomplete {
        background: $azure url('~Images/icons/action-null.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
        border: 1px solid red;
    }

    &--incomplete:hover {
        background: url('~Images/icons/check-selected.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
        border: none;
    }

    &--incomplete:focus {
        background: $azure url('~Images/icons/action-null.svg') no-repeat 50% 50% !important;
        background-size: 26px !important;
        border: none;
    }
    @extend .UserSelect--none;
}

.File__icon{
    align-self: center;

    @extend .UserSelect--none;
}



.File__input[type="text"]{
  height: 30px !important;
  background: $white !important;
  border-bottom: 0px solid $white !important;
  box-shadow: 0 4px 30px 0 rgba(0, 0, 0, 0.1) !important;

  margin: 0 20px 0 0 !important;

  flex: 1;

  &:focus{
    height: 30px !important;
    background: $white !important;
    border-bottom: 0px solid $white !important;
    box-shadow: 0 4px 30px 0 rgba(0, 0, 0, 0.4) !important;

  }
}

.File__container {
  position: relative;
  flex: 1;
}

.File__btn {
  &--round {
    height: 30px;
    width: 34px !important;
    padding: 0;

    border: 0 !important;

    border-radius: 15px;
  }

  &--add {
    background: url(~Images/icons/check-selected.svg) no-repeat center center;
    background-size: 30px;
    flex: 0 1 30px;

    &:hover {
      background: url(~Images/icons/check-selected.svg) no-repeat center center;
      background-size: 34px;
    }

    &:focus {
      background: url(~Images/icons/check-selected.svg) no-repeat center center;
      background-size: 30px;
    }
    &:disabled {
      background: url(~Images/icons/check-selected-disabled.svg) no-repeat center center !important;
      background-size: 30px !important;
    }
  }

  &--cancel {
    background: url(~Images/icons/close.svg) no-repeat center center;
    background-size: 26px;
    flex: 0 1 30px;

    &:hover {
      background: url(~Images/icons/close.svg) no-repeat center center;
      background-size: 30px;
    }

    &:focus {
      background: url(~Images/icons/close.svg) no-repeat center center;
      background-size: 26px;
    }
  }
}

.File__paragragh--external {
  background: url(~Images/icons/offsite-grey.svg) no-repeat right center;
  padding-right: 20px;
  background-size: 15px;
  &:hover {
    text-decoration: underline;
    background-size: 17px;
    padding-right: 23px;
  }
}