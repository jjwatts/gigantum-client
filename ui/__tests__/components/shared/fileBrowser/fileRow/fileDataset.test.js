// vendor
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import relayTestingUtils from '@gigantum/relay-testing-utils';
import path from 'path';
// data
import codeData from 'Tests/components/labbook/code/__relaydata__/CodeBrowser.json';
// components
import File from 'Components/shared/fileBrowser/fileRow/File';
import UserIdentity from 'JS/Auth/UserIdentity';

jest.mock('JS/Auth/UserIdentity', () => {
    return {
      getUserIdentity: jest.fn().mockResolvedValue({
        data: {
          userIdentity: {
            isSessionValid: true,
         },
       },
     }),
   }
 });


let edge = codeData.data.labbook.code.allFiles.edges[0];
edge.node.isDatasetRoot = true;
edge.node.isLocal = false;
let filename = path.basename(edge.node.key);
let file = {
  edge,
};

const mutationData = {
  owner: 'uitest',
  labbookName: 'ui-test-project',
  section: 'data',
  connection: 'CodeBrowser_allFiles',
  favoriteConnection: 'CodeFavorite_favorites',
  parentId: codeData.data.labbook.id,
};

let fixtures = {
  connectDragSource: jsx => jsx,
  closeLinkModal: jest.fn(),
  setState: jest.fn(),
  updateChildState: jest.fn(),
  codeDirUpload: jest.fn(),
  isSelected: true,
  filename,
  key: file.edge.node.key,
  multiSelect: false,
  mutationData,
  fileData: file,
  mutations: {
    moveLabbookFile: jest.fn(),
    downloadDatasetFiles: jest.fn(),
  },
  sort: 'all',
  reverse: true,
  childrenState: {},
  section: 'data',
  isDragging: false,
  isParent: true,
};


describe('Dataset File component', () => {
  it('File Snapshot', () => {
    const component = renderer.create(<File.DecoratedComponent {...fixtures}/>);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  const component = mount(<File.DecoratedComponent {...fixtures}/>);
});
