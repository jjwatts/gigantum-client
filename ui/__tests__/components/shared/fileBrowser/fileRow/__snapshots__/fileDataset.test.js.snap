// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset File component File Snapshot 1`] = `
<div
  className="File"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="File__row"
    style={
      Object {
        "paddingLeft": "NaNpx",
      }
    }
  >
    <button
      className="CheckboxMultiselect CheckboxMultiselect__check"
      onClick={[Function]}
      type="button"
    />
    <div
      className="File__cell File__cell--name"
    >
      <div
        className="File__icon null"
      />
      <div
        className="File__text"
      />
    </div>
    <div
      className="File__cell File__cell--edit hidden"
    >
      <div
        className="File__container"
      >
        <input
          className="File__input"
          draggable={true}
          onChange={[Function]}
          onClick={[Function]}
          onDragStart={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="folder"
        />
      </div>
      <div
        className="flex justify-space-around"
      >
        <button
          className="File__btn--round File__btn--cancel File__btn--rename-cancel"
          onClick={[Function]}
          type="button"
        />
        <button
          className="File__btn--round File__btn--add File__btn--rename-add"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
      </div>
    </div>
    <div
      className="File__cell File__cell--size"
    >
      0B
    </div>
    <div
      className="File__cell File__cell--date"
    >
      Fri Jan 18 2019
    </div>
    <div
      className="File__cell File__cell--menu"
    >
      <div
        className="ActionsMenu"
      >
        <button
          className="ActionsMenu__item Btn Btn--fileBrowser Tooltip-data Tooltip-data--small Btn--round Btn__addFolder visibility-hidden"
          data-click-id="addFolder"
          data-tooltip="Add Subfolder"
          disabled={false}
          onClick={[Function]}
          type="button"
        />
        <div
          className="relative"
        >
          <button
            className="ActionsMenu__item Btn Btn--fileBrowser Btn__delete-secondary Btn--round Tooltip-data Tooltip-data--small"
            data-tooltip="Must download before deleting"
            disabled={true}
            onClick={[Function]}
            type="button"
          />
          <div
            className="ActionsMenu__popup hidden Tooltip__message"
          >
            <div
              className="Tooltip__pointer"
            />
            <p>
              Are you sure?
            </p>
            <div
              className="flex justify--space-around"
            >
              <button
                className="File__btn--round File__btn--cancel"
                onClick={[Function]}
                type="button"
              />
              <button
                className="File__btn--round File__btn--add"
                onClick={[Function]}
                type="button"
              />
            </div>
          </div>
        </div>
        <button
          className="ActionsMenu__item Btn Btn--fileBrowser Btn__rename Tooltip-data Tooltip-data--small Btn--round"
          data-tooltip="Must download before renaming"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
      </div>
      <div
        className="DatasetActionsMenu"
      >
        <button
          className="Btn Btn__FileBrowserAction DatasetActionsMenu__item Btn__FileBrowserAction--download"
          onClick={[Function]}
          type="button"
        >
          Download
        </button>
      </div>
    </div>
  </div>
</div>
`;
